third_party/amd/lib/TritonAMDGPUToLLVM/ElementwiseOpToLLVM.cpp

```diff
@@ -1370,6 +1371,146 @@ struct RsqrtOpConversion
   bool ftz;
 };
 
+static inline std::pair<Value, Value>
+ScaleUpIfDenorm(ConversionPatternRewriter &rewriter, Location loc,
+                const Value &src, float scaleThreshold, float scaleFactor) {
+  Value needScale = fcmp_ogt(f32_val(scaleThreshold), src);
+  Value scaledSrc = fmul(f32_ty, src, f32_val(scaleFactor));
+  Value selectedSrc = select(needScale, scaledSrc, src);
+  return {needScale, selectedSrc};
+}
+
+static inline Value ScaleDownIfDenorm(ConversionPatternRewriter &rewriter,
+                                      Location loc, const Value &src,
+                                      Value needScale, float scaleFactor) {
+  Value scaledSrc = fmul(f32_ty, src, f32_val(scaleFactor));
+  return select(needScale, scaledSrc, src);
+}
+
+struct SqrtOpConversion
+    : ElementwiseOpConversionBase<mlir::math::SqrtOp, SqrtOpConversion> {
+  using ElementwiseOpConversionBase<
+      mlir::math::SqrtOp, SqrtOpConversion>::ElementwiseOpConversionBase;
+
+  explicit SqrtOpConversion(LLVMTypeConverter &typeConverter,
+                            ModuleAxisInfoAnalysis &axisInfoAnalysis, bool ftz,
+                            PatternBenefit benefit)
+      : ElementwiseOpConversionBase(typeConverter, axisInfoAnalysis, benefit),
+        ftz(ftz) {}
+
+  SmallVector<Value> createDestOps(mlir::math::SqrtOp op, OpAdaptor adaptor,
+                                   ConversionPatternRewriter &rewriter,
+                                   Type elemTy, MultipleOperandsRange operands,
+                                   Location loc) const {
+    // Only handle f32 here. Other data types will be lowered to LLVM::SqrtOp by
+    // MLIR.
+    if (elemTy.getIntOrFloatBitWidth() != 32)
+      return {};
+
+    Value needScale = false_val();
+    Value scaledSrc = operands[0][0];
+    if (!ftz) {
+      // In case of non-ftz, we need to scale up the input by a factor 2^{32},
+      // if it's below 2^{-96}, to prevent it from being flushed by
+      // llvm.amdgcn.sqrt.f32.
+      //
+      // And we will scale down the result afterwards to get correct result.
+      // Reference:
+      // https://github.com/llvm/llvm-project/blob/0876c11ceeb093904decc4d89bef213d483a5656/llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp#L5235-L5314.
```

Comment:
Could you add explicit comments to explain what simplifications we have here comparing to the llvm version? I.e., why not use the LLVM version. Similarly for the following PreciseSqrt case.